### PR TITLE
Fixed bugs that prevented consecutive writes to HTTP response from working properly

### PR DESCRIPTION
**Pull request by bainian12345**
- Added lock to ClientWebSocket to prevent exception from the websocket API when multiple sends are attempted simultaneously.

- Added trimming to HTTP response buffer to ensure that empty bytes do not get sent.

- Added tests to ensure consecutive writes to HTTP response is working properly.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 	<groupId>com.microsoft.azure</groupId>
 	<artifactId>azure-relay</artifactId>
-	<version>0.0.1-PREVIEW</version>
+	<version>0.0.2-PREVIEW</version>
```
Comment (dlstucki) [inline pom.xml]: >.2 [](start = 13, length = 2)

Are you just about to create a new release? That's typically when we increase the version--as a last step before publishing.

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 2]
```diff
@@ -36,6 +36,7 @@ public ClientWebSocket(TrackingContext trackingContext, AutoShutdownScheduledExe
 		this.fragmentQueue = new InputQueue<MessageFragment>(this.executor);
 		this.closeReason = null;
 		this.trackingContext = trackingContext;
+		this.socketLock = new AsyncLock(HybridConnectionListener.EXECUTOR);
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: >HybridConnectionListener [](start = 34, length = 24)

I thought we ran into deadlock issues when the send side used the same thread pool as the receiver side. ClientWebSocket is used by both, correct?

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 3]
```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
...
-							RelayLogger.logEvent("writingBytes", this, "binary");
-							if (data instanceof byte[]) {
-								bytes = ByteBuffer.wrap((byte[]) data);
-							} 
-							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
+				RemoteEndpoint.Basic remote = this.session.getBasicRemote();
+				RelayLogger.logEvent("writingBytes", this, mode.toString());
+				
+				// The websocket API will throw if multiple sends are attempted on the same websocket simultaneously
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: >The websocket API will throw if multiple sends are attempted on the same websocket simultaneously [](start = 7, length = 97)

I'm curious if it's the libraries job to ensure users don't do this bad thing or the user's responsibility. My inclination is that if we try to do it for them then they don't get any guarantees as to the order of their data. If we don't do it for them, then they must wait until the previous write finishes before making the next write and their data is actually guaranteed to be in the right order.

Basically this doesn't guarantee ordering:
webSocket.writeAsync( ...) ; // bytes 0-9, NO WAIT
webSocket.writeAsync(...); // bytes 10-19, NO WAIT

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 4]
```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
...
+								else if (data instanceof ByteBuffer) {
+									bytes = (ByteBuffer) data;
+								}
+								else if (data instanceof String){
+									bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+								}
+								
+								int bytesToSend = bytes.remaining();
+								remote.sendBinary(bytes, isEnd);
+								RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(bytesToSend));
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: >String.valueOf(bytesToSend) [](start = 55, length = 27)

Does this log the actual customer data? We shouldn't be doing that without them turning it on _very_ explicitly.

File: src/main/java/com/microsoft/azure/relay/RelayLogger.java [hunk 5]
```diff
@@ -128,6 +128,7 @@ private static void init() {
 		map.put("httpSendResponse", "%s: Sending the response command on the %s connection, status: %s.");
 		map.put("httpSendResponseFinished", "%s: Finished sending the response command on the %s connection, status: %s.");
 		map.put("httpUserRequestHandlerException", "%s: exception in the user's request handler.");
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/RelayLogger.java]: >exception in the user's request handler. [](start = 50, length = 40)

I know you're not touching this line but I happened to look at it anyway...

In the case of user request handler error do no details of the exception get logged?

If something more is needed feel free to file an issue and resolve this comment.

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java [hunk 6]
```diff
@@ -209,24 +209,27 @@ void invokeRequestHandler(RequestCommandAndStream requestAndStream) {
...
 			ListenerCommand listenerCommand = new ListenerCommand(null);
 			listenerCommand.setResponse(responseCommand);
 			String command = listenerCommand.getResponse().toJsonString();
 
 			// We need to respond over the rendezvous connection
-			CompletableFuture<Void> sendCommandTask = this.rendezvousWebSocket.writeAsync(command, timeout, WriteMode.TEXT)
-					.thenAccept(bytesWritten -> {
-						RelayLogger.logEvent("httpSendResponseFinished", this, "rendezvous", String.valueOf(responseCommand.getStatusCode()));
-					});
+			CompletableFuture<Void> sendCommandTask = this.ensureRendezvousAsync(timeRemaining.remainingTime())
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: >ensureRendezvousAsync [](start = 50, length = 21)

I wonder if the fact that this else block is only hit when `this.rendezvousWebSocket != null` means ensureRendezvousAsync can be skipped?

Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: It appears my statement is not true. Since ensureRendezvousAsync assigns to this.rendezvousWebSocket before it is connected, and not after what I had hoped to be the case won't work.

---
In reply to: [291718830](https://github.com/Azure/azure-relay-java/pull/45#discussion_r291718830) [](ancestors = 291718830)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 7]
```diff
@@ -293,4 +311,19 @@ private static void httpRequestSender(String method, String msgExpected, String
...
 		assertEquals("Http connection sender did not receive the expected response message.", msgExpected, builder.toString());
 	}
+	
+	private static void httpRequestHandlerMultipleSend(RelayedHttpListenerContext context, String firstMsg, String secondMsg) {
+		RelayedHttpListenerResponse response = context.getResponse();
+		response.setStatusCode(STATUS_CODE);
+		response.setStatusDescription(STATUS_DESCRIPTION);
+		
+		try {
+			response.getOutputStream().write(firstMsg.getBytes());
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >write [](start = 30, length = 5)

Does write call writeAsync and wait for it to complete or just kick it off?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 8]
```diff
@@ -204,6 +204,24 @@ public void httpLargePOSTAndLargeResponseTest() throws IOException {
         httpRequestSender("POST", largeStr, largeStr);
 	}
 	
+	@Test
+	public void httpWriteSmallThenSmallResponseTest() throws IOException {
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >httpWriteSmallThenSmallResponseTest [](start = 13, length = 35)

Does anything in the test suite fail without the send async lock that's being introduced?

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java (hunk 3 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: Or is it to solve the issue where user calls writeAsync the closeAsync without waiting?

CF writeFuture = ws.writeAsync(...);
CF closeFuture = ws.closeAsync(); // close waits for inflight send to drain?

---
In reply to: [291712180](https://github.com/Azure/azure-relay-java/pull/45#discussion_r291712180) [](ancestors = 291712180)

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 9]
```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
...
-							} 
-							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
+				RemoteEndpoint.Basic remote = this.session.getBasicRemote();
+				RelayLogger.logEvent("writingBytes", this, mode.toString());
+				
+				// The websocket API will throw if multiple sends are attempted on the same websocket simultaneously
+				// Therefore a lock is used on the send operation
+				return this.socketLock.acquireThenCompose(timeout, () -> {
+					return CompletableFutureUtil.timedRunAsync(timeout, () -> {
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: >timeout [](start = 48, length = 7)

I suppose in a case like this a TimeoutHelper should be used. Time waiting for the socketLock should be deducted from the remaining operation time.

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 10]
```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
...
+								ByteBuffer bytes = null;
+								if (data instanceof byte[]) {
+									bytes = ByteBuffer.wrap((byte[]) data);
+								} 
+								else if (data instanceof ByteBuffer) {
+									bytes = (ByteBuffer) data;
+								}
+								else if (data instanceof String){
+									bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+								}
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: Isn't it possible for an Object to not fall into any of these if/else-if combinations? Ideally this API should throw instead of doing an empty send in that case.

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java [hunk 11]
```diff
@@ -324,13 +327,20 @@ public void setWriteTimeout(Duration writeTimeout) {
...
-							.sendBytesOverRendezvousAsync(this.writeBufferStream, timeout).thenRun(() -> {
-								this.writeBufferStream.clear();
-								if (this.writeBufferFlushTimer != null) {
-									this.writeBufferFlushTimer.cancel();
-								}
-							}));
+					return sendResponseTask.thenCompose($void -> {
+						// Get a new buffer backed by the non empty segment of the write buffer array
+						this.writeBufferStream.flip();
+						ByteBuffer trimmedBuffer = ByteBuffer.wrap(
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: >ByteBuffer.wrap( [](start = 33, length = 16)

What benefit does wrapping the ByteBuffer add here?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 8 above)
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >httpWriteSmallThenSmallResponseTest [](start = 13, length = 35)

Would anything in the tests catch and fail on a scenario where the message arrived with extra '\0' characters? (Converting from bytes to String might throw away trailing '\0' chars...

"FirstSecond\0\0\0\0"

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java [hunk 12]
```diff
@@ -324,13 +327,20 @@ public void setWriteTimeout(Duration writeTimeout) {
...
+								this.writeBufferStream.array(), 
+								this.writeBufferStream.position(), 
+								this.writeBufferStream.remaining());
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: Can you help me understand this? Why is a clone of the writeBufferStream needed (the wrap which I believe creates something basically equivalent)?

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java [hunk 13]
```diff
@@ -324,13 +327,20 @@ public void setWriteTimeout(Duration writeTimeout) {
...
+						return this.connection.sendBytesOverRendezvousAsync(trimmedBuffer, timeout);
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: >timeout [](start = 73, length = 7)

Should this be using TimeoutHelper? Is it allowed to take 2*timeout?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 8 above)
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >httpWriteSmallThenSmallResponseTest [](start = 13, length = 35)

Are there any tests which write some bytes to the response, pausing allowing the flush timer to send those bytes, then send some more bytes in the response body and validate the exact number of bytes and values are transmitted to the client?

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java [hunk 14]
```diff
@@ -209,24 +209,27 @@ void invokeRequestHandler(RequestCommandAndStream requestAndStream) {
 			return this.listener.sendControlCommandAndStreamAsync(listenerCommand, responseBodyBuffer, timeout)
 					.thenRun(() -> RelayLogger.logEvent("httpSendResponseFinished", this, "control", String.valueOf(responseCommand.getStatusCode())));
 		} else {
+			TimeoutHelper timeRemaining = new TimeoutHelper(timeout);
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: >TimeoutHelper timeRemaining = new TimeoutHelper(timeout); [](start = 3, length = 57)

I wouldn't bother changing this now, but keep in mind "timeoutHelper" is really a replacement for "timeout" argument. Any code path which would use timeout should really be aware of timeoutHelper. My personal guideline is therefore to try to keep TimeoutHelper at the top of the method.

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java (hunk 12 above)
Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: It's to wrap only the part of the buffer that has been written to, so the empty bytes in the buffer aren't sent.

---
In reply to: [292073271](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292073271) [](ancestors = 292073271)

Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: Does that actually happen without the wrap here? 

---
In reply to: [292098393](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292098393) [](ancestors = 292098393,292073271)

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java (hunk 3 above)
Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: This lock is to guard the websocket from multiple writes concurrently, not related to close.

All the unit tests are doing synchronized writes as of now and not running into this issue. I will do more investigation on this later in a separate PR. Probably going to leave this lock out for now.

---
In reply to: [292064317](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292064317) [](ancestors = 292064317,291712180)

Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: If a user calls writeAsync twice concurrently I think that ideally we should give them an exception. Even if no exception it's pretty much into undefined territory.

**However**, if our timer based flush is racing with the user calling another writeAsync we almost certainly need to serialize things there.

---
In reply to: [292622840](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292622840) [](ancestors = 292622840,292064317,291712180)

File: src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java (hunk 12 above)
Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java]: You are right, what I needed was the flip() on the buffer, not wrapping it.

---
In reply to: [292098796](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292098796) [](ancestors = 292098796,292098393,292073271)

File: src/main/java/com/microsoft/azure/relay/RelayLogger.java (hunk 5 above)
Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/RelayLogger.java]: Should we log any details about user implementation? I was concerned about exposing user internal details

---
In reply to: [291715446](https://github.com/Azure/azure-relay-java/pull/45#discussion_r291715446) [](ancestors = 291715446)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 8 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Runs fine for now with synchronized writes, will do async writes in separate PR

---
In reply to: [292063905](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292063905) [](ancestors = 292063905)

File: src/main/java/com/microsoft/azure/relay/RelayLogger.java (hunk 5 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/RelayLogger.java]: It's far less common to have sensitive information in Exceptions themselves. Personally, I'd to the exception type, message, callstack in the tracing. Also, these traces go wherever our customers configure them to go after they opt-in by turning on tracing. Typically that's something that stays on the same machine as the user code. One has to be more careful when sending details over the wire or saving logs in multitenant systems or in the cloud.

---
In reply to: [292631454](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292631454) [](ancestors = 292631454,291715446)
